From 6d0cc8ae64546a27db25381fdfe0e79518ddfc85 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:06:54 -0400
Subject: [tools] perf/ui/tui: Fix off-by-one in hist_browser__update_nr_entries()

Message-id: <1410181752-25631-170-git-send-email-jolsa@redhat.com>
Patchwork-id: 91862
O-Subject: [PATCH RHEL7.1 BZ1134356 169/307] perf ui/tui: Fix off-by-one in hist_browser__update_nr_entries()
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit c481f9301183260a78e55fa4d70d977b68c81846
Author: Namhyung Kim <namhyung@kernel.org>
Date: Tue Apr 22 13:56:11 2014 +0900

description
===========
The nr_entries variable is increased inside the loop in the function
but it always count the first entry regardless of it's filtered or
not; caused an off-by-one error.

It'd become a problem especially there's no entry at all - it'd get a
segfault during referencing a NULL pointer.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/ui/browsers/hists.c b/tools/perf/ui/browsers/hists.c
index 4d41698..311226e 100644
--- a/tools/perf/ui/browsers/hists.c
+++ b/tools/perf/ui/browsers/hists.c
@@ -1348,10 +1348,10 @@ static void hist_browser__update_pcnt_entries(struct hist_browser *hb)
  u64 nr_entries = 0;
  struct rb_node *nd = rb_first(&hb->hists->entries);
 
- while (nd) {
+ while ((nd = hists__filter_entries(nd, hb->hists,
+        hb->min_pcnt)) != NULL) {
   nr_entries++;
-  nd = hists__filter_entries(rb_next(nd), hb->hists,
-        hb->min_pcnt);
+  nd = rb_next(nd);
  }
 
  hb->nr_pcnt_entries = nr_entries;
-- 
1.7.1